### PR TITLE
Move level properties into their own type

### PR DESCRIPTION
**Pull request by bencodeorg**
Motivation for this change discussed in this Slack thread:

https://codedotorg.slack.com/archives/C044AGTKW3D/p1690842845194209

Update: (Sanchit) picking up this work while Ben is out. This change adds the entire `levelProperties` object into the lab2 redux store, to be used by AI Chat. Note that properties derived from levelProperties like `isProjectLevel`, `hideShareAndRemix`, etc still exist on the redux store as separate properties, so that we don't have to also update all the components that reference these fields. However as a follow-up, we should update those components to access those fields via levelProperties, and remove the fields from the store.

## Testing story

Tested locally; was able to retrieve AI Chat specific properties in AiChatView, and confirmed that Music Lab still works on standalone, level, script level, and project beats pages.

### CODE REVIEW

File: apps/src/lab2/lab2Redux.ts [hunk 1]
```diff
@@ -227,10 +226,10 @@ const labSlice = createSlice({
       state.channel = action.payload;
     },
     setHideShareAndRemix(state, action: PayloadAction<boolean>) {
```
Comment (sanchitmalhotra126) [inline apps/src/lab2/lab2Redux.ts]: We should also remove these setters (setHideAndShareRemix, setIsProjectLevel) and just set levelProperties directly. We can actually probably just do that in `onLevelChange` - instead of passing in `levelData` and `appName`, we can just pass in `levelProperties` and set it on the state object.

File: apps/src/lab2/types.ts [hunk 2]
```diff
@@ -81,7 +81,7 @@ export interface Level {
   isAquaticLevel: boolean;
 }
 
-export interface LevelProperties {
+export interface ServerLevelProperties {
```
Comment (sanchitmalhotra126) [inline apps/src/lab2/types.ts]: Hmm, this is interesting...trying to think if makes sense to have this as a separate type, or just use the `LevelProperties` type throughout. An advantage with this is that we can narrow the boolean values to actual boolean types, but i'm trying to think if it's worth maintaining two different types for that. @molly-moen curious what you think?

**Review by sanchitmalhotra126 — COMMENTED** [review]
Thanks for taking a crack at this! I think we'll need to make a few more updates to make lab2 labs work correctly with these changes - any components that reference any of the level properties fields (isProjectLevel, hideShareAndRemix, levelData, appName) will need to be updated to access them via `levelProperties.<property>`, i.e. (`state.lab.isProjectLevel` => `state.lab.levelProperties.isProjectLevel`). It should just be a matter of updating selector functions in a few components.

Alternatively, if it's easier, an intermediate option could be to keep all the sub-properties on the state object (isProjectLevel, etc), and then in `onLevelChange`, we can set each of the sub-properties based from levelProperties in addition to the full levelProperties object, for example
```
onLevelChange(state, action ...) {
   state.isProjectLevel = convertOptionalStringToBoolean(action.payload.levelProperties.isProjectLevel, false);
   state.hideAndShareRemix = convertOptionalStringToBoolean(action.payload.levelProperties.hideShareAndRemix, false);
   state.levelData = action.payload.levelProperties.levelData;
   state.appName = action.payload.levelProperties.appName;
   state.levelProperties = action.payload.levelProperties
```
This way at least you wouldn't have to change any other components right now, and we can convert over one field at a time

File: apps/src/lab2/lab2Redux.ts [hunk 3]
```diff
@@ -14,9 +14,9 @@ import {
   AppName,
   Channel,
   LevelData,
-  LevelProperties,
   ProjectManagerStorageType,
   ProjectSources,
+  LevelProperties,
```
Comment (fisher-alice) [inline apps/src/lab2/lab2Redux.ts:19]: Nit: why did you move this (out of alpha order)?

File: apps/src/lab2/responseValidators.ts [hunk 4]
```diff
@@ -1,26 +1,45 @@
...
+  for (const key of Object.keys(response)) {
+    if (response[key] === 'true') {
+      response[key] = true;
+    }
+    if (response[key] === 'false') {
+      response[key] = false;
+    }
+  }
+
+  return response as unknown as LevelProperties;
```
Comment (fisher-alice) [inline apps/src/lab2/responseValidators.ts:42]: Could you explain this syntax `return response as unknown as LevelProperties`?

File: apps/src/lab2/responseValidators.ts [hunk 5]
```diff
@@ -1,26 +1,45 @@
...
   }
   if (blocklySource.blocks === undefined) {
     throw new ValidationError('Missing required field: blocks');
   }
 
-  return projectSources;
+  return response as unknown as ProjectSources;
+};
+
+export const LevelPropertiesValidator: ResponseValidator<
```
Comment (fisher-alice) [inline apps/src/lab2/responseValidators.ts:25]: Do you recommend that we add a `ChatCompletionValidator` for the response returned from the Open AI chat completion controller? Just curious.

File: apps/src/lab2/lab2Redux.ts [hunk 6]
```diff
@@ -53,13 +53,16 @@ export interface LabState {
...
-  // Level data for the current level
-  levelData: LevelData | undefined;
-  hideShareAndRemix: boolean;
-  isProjectLevel: boolean;
   // Validation status for the current level. This is used by the progress system to determine
   // what instructions to display and if the user has satisfied the validation conditions, if present.
   validationState: ValidationState;
+  // Level properties for the current level.
+  levelProperties: LevelProperties | undefined;
+  // Note: the following properties are derived from levelProperties. We should update components that consume
```
Comment (fisher-alice) [inline apps/src/lab2/lab2Redux.ts:61]: And when components are updated to use `levelProperties`,  the fields below can be removed?

File: apps/src/util/HttpClient.ts [hunk 7]
```diff
@@ -4,7 +4,7 @@ import {
 } from './AuthenticityTokenStore';
 
 export type ResponseValidator<ResponseType> = (
-  bodyJson: object
+  bodyJson: Record<string, unknown>
```
Comment (fisher-alice) [inline apps/src/util/HttpClient.ts:7]: Would you mind explaining why you use `unknown` vs `any` for value type? Is `unknown` always preferred?

**Review by fisher-alice — APPROVED** [review]
Learning a lot from this PR -I am new to this area of code base so asking lots of questions. LGTM but I know you left a question for @molly-moen.

File: apps/src/lab2/types.ts (hunk 2 above)
Comment (molly-moen) [reply apps/src/lab2/types.ts]: looks like you removed this in favor of the validator, that seems like the right call!

**Review by molly-moen — APPROVED** [review]
Looks good to me!

File: apps/src/lab2/lab2Redux.ts (hunk 3 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/lab2Redux.ts:19]: Oh good catch, no particular reason :) I think I must have removed the import then re-added it

File: apps/src/lab2/responseValidators.ts (hunk 4 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/responseValidators.ts:42]: Yeah, this is a bit confusing 🙂 `response` is of type `Record<string, unknown>` and TypeScript doesn't believe that a `Record<string, unknown>` type can be assigned to `LevelProperties` because there's no overlap in the types. However, we do know that this is response is indeed a `LevelProperties` type because we know which API we're calling, and it's passed validation. So in order to type cast it, we need to first cast it to `unknown` which removes all typing, and then cast it to `LevelProperties`. More on the `unknown` type here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type. Basically it's a way to say that this type could be anything, but we have to first narrow the type (by casting it here) before treating it as a specific type.

File: apps/src/lab2/responseValidators.ts (hunk 5 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/responseValidators.ts:25]: We could, yeah. It's not necessary to call the API, but if we want to be stricter about the type we're getting back from the server it could be good to have. That being said, since we make a POST request, we'll probably have to update the HttpClient to accept a validator since it doesn't currently for post requests. Alternatively, we can just perform our own validation after we get the response back.

File: apps/src/lab2/lab2Redux.ts (hunk 6 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/lab2Redux.ts:61]: Yep!

File: apps/src/util/HttpClient.ts (hunk 7 above)
Comment (sanchitmalhotra126) [reply apps/src/util/HttpClient.ts:7]: Yep, `unknown` is basically a safer version of `any`. The `any` type is really powerful because it removes any typing and allows you to call any function or reference any property that may or may not exist. It can often serve an escape hatch to turn typed code back into untyped vanilla JS, so most linting rules try to limit its use. `unknown` is similar but you can't call any functions or reference any properties on it, until you narrow it to a more specific type. So it's a way to say "this could be anything, but we don't know what it is until we explicitly tell TypeScript what type it is".

For example -
```
const myAny: any = 'string';
myAny.doSomething(); // this will compile even though the function doesn't exist. at runtime this will throw an error

const myUnknown: unknown = 'string';
myUnknown.doSomething(); // this will not compile because we don't know what properties are on this type
(myUnknown as string).charAt(0); // this will now compile because we've narrowed the type to string
```

File: apps/src/lab2/responseValidators.ts (hunk 4 above)
Comment (fisher-alice) [reply apps/src/lab2/responseValidators.ts:42]: Interesting - thanks!
